### PR TITLE
[CARE-1802] Refactor - Replace usages of LoaderExtension with Placeholders

### LINKED ISSUE #500: Get rid of Loader component
The Placeholders functionality supersedes Loaders.

### PR DESCRIPTION
**Pull request by e1himself**
Fixes #500 

- Replace `LoaderExtension` with Placeholders
- Replace `ImageCandidateElement` functionality with Placeholders

Result: minus ~1000~ 600 LOC :tada:

### CODE REVIEW

File: packages/slate-editor/src/extensions/galleries/GalleriesExtension.tsx [hunk 1]
```diff
@@ -17,16 +19,27 @@ import {
...
+    onEdited?: (
+        editor: Editor,
+        gallery: GalleryNode,
+        extra: {
+            successfulUploads: number;
+            failedUploads: Error[];
+        },
+    ) => void;
+    onShuffled?: (editor: Editor, updated: GalleryNode, original: GalleryNode) => void;
+    withMediaGalleryTab: false | { enabled: true; newsroom: NewsroomRef };
```
Comment (MohammadxAli) [inline packages/slate-editor/src/extensions/galleries/GalleriesExtension.tsx:31]: Isn't `enabled: boolean` more appropriate than `enabled: true`? I'm currently in the process of testing features in Prezly by integrating the changes from this PR, and this error appears:

```tsx
<Editor
  withGalleries={{
    withMediaGalleryTab: { newsroom: story.newsroom, enabled: isGalleryEnabled },
  }}
  // ...
/>
```

```
Type 'boolean' is not assignable to type 'true'.ts(2322)
```

File: packages/slate-editor/src/extensions/galleries/components/GalleryElement.tsx [hunk 2]
```diff
@@ -1,36 +1,133 @@
...
+            }
+
+            return awaitUploads(filePromises);
+        }
+
+        setUploading(true);
+
+        try {
+            const result = await upload();
+
```
Comment (MohammadxAli) [inline packages/slate-editor/src/extensions/galleries/components/GalleryElement.tsx:85]: Shouldn't `setUploading(true)` live in the `try` block?

File: packages/slate-editor/src/modules/editor/getEnabledExtensions.ts [hunk 3]
```diff
@@ -210,9 +200,30 @@ export function* getEnabledExtensions(parameters: Parameters): Generator<Extensi
...
+                });
+            },
+        });
+
         yield FileAttachmentExtension({
-            onEdit: handleEditAttachment,
-            onRemove: handleRemoveAttachment,
+            onEdited(editor, updated) {
+                // TODO: It seems it would be more useful to only provide the changeset patch in the event payload.
+                EventsEditor.dispatchEvent(editor, 'attachment-edited', {
```
Comment (MohammadxAli) [inline packages/slate-editor/src/modules/editor/getEnabledExtensions.ts:215]: 👀

File: packages/slate-editor/src/modules/nodes-hierarchy/tests/Image.test.tsx [hunk 4]
```diff
@@ -9,7 +9,7 @@ describe('nodes-hierarchy / Image', () => {
     it('should be kept after normalization', function () {
         const editor = (
             <editor>
-                <h:image-candidate href="image-candidate-href" src="image-candidate-src" />
+                {/*<h:image-candidate href="image-candidate-href" src="image-candidate-src" />*/}
```
Comment (MohammadxAli) [inline packages/slate-editor/src/modules/nodes-hierarchy/tests/Image.test.tsx:12]: 👀

File: packages/slate-editor/src/modules/editor/Editor.tsx [hunk 5]
```diff
@@ -121,7 +119,10 @@ export const Editor = forwardRef<EditorRef, EditorProps>((props, forwardedRef) =
 
     const events = useMemo(() => new Events<EditorEventMap>(), []);
     const containerRef = useRef<HTMLDivElement>(null);
-    const { onOperationEnd, onOperationStart } = usePendingOperation(onIsOperationPendingChange);
+
+    // TODO: Wire `onOperationStart` and `onOperationEnd` to the Placeholder extension
+    // const { onOperationEnd, onOperationStart } = usePendingOperation(onIsOperationPendingChange);
```
Comment (MohammadxAli) [inline packages/slate-editor/src/modules/editor/Editor.tsx:124]: 👀

File: packages/slate-editor/src/extensions/placeholders/elements/ImagePlaceholderElement.tsx [hunk 6]
```diff
@@ -51,6 +51,11 @@ export function ImagePlaceholderElement({
 
         images.forEach((filePromise, i) => {
             const uploading = toProgressPromise(filePromise).then((fileInfo: PrezlyFileInfo) => {
+                if (!fileInfo.isImage) {
+                    // FIXME: What if it's not an image?
+                    //        It would make sense to re-route this to an Attachment node.
```
Comment (MohammadxAli) [inline packages/slate-editor/src/extensions/placeholders/elements/ImagePlaceholderElement.tsx:56]: 👀

File: packages/slate-editor/src/components/EditorBlock/EditorBlock.module.scss [hunk 7]
```diff
@@ -70,6 +70,33 @@ $editor-block-outline-width: 6px;
     }
 }
 
+.LoadingIndicator {
+    display: block;
+    position: absolute;
+    left: 0;
+    top: 0;
```
Comment (MohammadxAli) [inline packages/slate-editor/src/components/EditorBlock/EditorBlock.module.scss:77]: _Comment in an unrelated place_

Here's a potential bug:


https://github.com/prezly/slate/assets/67554982/f8f9b429-fcf4-4439-8328-8c9b0f504b97


Steps to reproduce:

1. Add a gallery
2. Add some images
3. Click on the "Edit"
4. Close the modal
5. It gets stuck in the loading state

But it shouldn't.

File: packages/slate-editor/src/extensions/galleries/components/GalleryElement.tsx (hunk 2 above)
Comment (e1himself) [reply packages/slate-editor/src/extensions/galleries/components/GalleryElement.tsx:85]: Why? It cannot fail AFAIK

File: packages/slate-editor/src/extensions/galleries/GalleriesExtension.tsx (hunk 1 above)
Comment (e1himself) [reply packages/slate-editor/src/extensions/galleries/GalleriesExtension.tsx]: 8b0ade2fed84c27c222fdbaee6d5c3e892b224ea

File: packages/slate-editor/src/modules/nodes-hierarchy/tests/Image.test.tsx (hunk 4 above)
Comment (e1himself) [reply packages/slate-editor/src/modules/nodes-hierarchy/tests/Image.test.tsx]: ec3e1c7064c44384eb7e20db9fd676c9d9f23fff

File: packages/slate-editor/src/components/EditorBlock/EditorBlock.module.scss (hunk 7 above)
Comment (e1himself) [reply packages/slate-editor/src/components/EditorBlock/EditorBlock.module.scss:77]: Great catch! Thanks!

File: packages/slate-editor/src/extensions/placeholders/elements/ImagePlaceholderElement.tsx (hunk 6 above)
Comment (e1himself) [reply packages/slate-editor/src/extensions/placeholders/elements/ImagePlaceholderElement.tsx]: 559b1b5746706406f025ec241d8d1cb8f5c91789

File: packages/slate-editor/src/components/EditorBlock/EditorBlock.module.scss (hunk 7 above)
Comment (e1himself) [reply packages/slate-editor/src/components/EditorBlock/EditorBlock.module.scss:77]: dbe8783fadb469eaf6925630a56cfe4a99a98f85

**Review by MohammadxAli — APPROVED** [review]
Couldn't find any regressions. Great stuff, well done! 🎉

File: packages/slate-editor/src/extensions/paste-images/lib/withImagesPasting.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+export function withImagesPasting({ onImagesPasted = noop }: Parameters = {}) {
+    return <T extends Editor>(editor: T): T => {
+        const parent = {
+            insertData: editor.insertData,
+        };
+
+        editor.insertData = (dataTransfer: DataTransfer) => {
+            if (!isFilesOnlyDataTransfer(dataTransfer)) {
+                // Handle images, if the pasted content is containing files only.
```
Comment (kudlajz) [inline packages/slate-editor/src/extensions/paste-images/lib/withImagesPasting.ts:25]: This comment seems to be incorrect? As this branch will only run if the data transfer is NOT files only.

**Review by kudlajz — APPROVED** [review]

File: packages/slate-editor/src/extensions/paste-images/lib/withImagesPasting.ts (hunk 8 above)
Comment (e1himself) [reply packages/slate-editor/src/extensions/paste-images/lib/withImagesPasting.ts:25]: It is correct. As images are handled below this `if`. But I see why you're confused.

How about this?

> Exit, if the pasted content is not containing files only.

Comment (e1himself) [reply packages/slate-editor/src/extensions/paste-images/lib/withImagesPasting.ts]: 31ec5e9066b3c03c5f9935d9f15a3094414f064c

Comment (kudlajz) [reply packages/slate-editor/src/extensions/paste-images/lib/withImagesPasting.ts]: I think that's clearer 👍

**Pull request merged**